### PR TITLE
Detect circular sym links

### LINKED ISSUE #764: Out of memory with symlinks
While trying to reproduce my stack overflow issue, stumbled upon this one:
```
$ mkdir foo && cd foo
$ ln -s . x
$ ln -s . y
$ jbuilder build
Fatal error: out of memory.
```
the OOM happening after a while; couldn't find a simple repro case for the stack overflow, which happened with a combination of many files and symlinks, but they seem related.

### PR DESCRIPTION
**Pull request by rgrinberg**
Fix #764

@diml any way to add tests for this? I tried adding Louis' example but dune errors out too early when it sees the symlink before actually running any tests.

### CODE REVIEW

**Comment by diml** [comment]
you can create the symlink in the test

File: src/file_tree.ml [hunk 1]
```diff
@@ -62,7 +62,16 @@ let ignore_file fn ~is_directory =
   (fn.[0] = '.' && fn.[1] = '#')
 
 let load ?(extra_ignored_subtrees=Path.Set.empty) path =
-  let rec walk path ~ignored : Dir.t =
+  let rec walk seen_real_paths path ~ignored : Dir.t =
+    let realpath =
+      try Path.of_string (Unix.readlink (Path.to_string path))
```
Comment (diml) [inline src/file_tree.ml]: shouldn't this be `Path.relative real_path (Unix.readlink ...))`

**Comment by diml** [comment]
I haven't thought about this in details, so here are a few questions:
- what happens when the target of the symlink is a symlink itself?
- what happens when the target of the symlink is an absolute path? We don't normalize external paths, so it's not clear to me whether the check will work or not
- what happens when the target of the symlink is a relative path but points outside of the workspace? This case should be supported

**Comment by diml** [comment]
> Tough one. Perhaps we need a maximum symlink depth to address this?

That seems fine. That's basically what all system calls taking filenames do

> Could you elaborate on this? The only issue I see here is that an external path could point to a path in the build dir or source tree which would cause problems for the symbolic link branch. But I think my PR addresses this.

I need to think a bit more about this, but I think there might be an issue. For instance, if `src` is a symlink to `/foo/bar` and then `/foo/bar` contains a symlink `x` pointing to `../baz`, then I think something odd might happen.

> This seems to work

What happens if the root contains a symlink pointing to `../../x`? I expect that `Path.relative` will fail

**Comment by rgrinberg** [comment]
> That seems fine. That's basically what all system calls taking filenames do

Ok, this is implemented.

> What happens if the root contains a symlink pointing to ../../x? I expect that Path.relative will fail

Let me create a proper test case for this.

**Comment by rgrinberg** [comment]
@diml I pushed a test case that seems to show that relative symlinks outside the workspace should work. Is this not triggering Path.relative as you'd expect?

File: src/stdune/path.ml [hunk 2]
```diff
@@ -521,3 +521,35 @@ let extension = Filename.extension
...
+      if n = 0 then
+        Result.Error `Maximum_depth_exceeded
+      else
+        match readlink fn with
+        | None -> Result.Ok fn
+        | Some p ->
+          if p = fn then
+            Result.Error `Cycle_detected
+          else
+            follow (pred n) p
```
Comment (diml) [inline src/stdune/path.ml]: I think there is a `Filename.concat` missing here

**Comment by diml** [comment]
I added a more complex test in 3a2e136e5ca716fdc9c65e1f189bb41414dcb4e3 which demonstrate what I was thinking about. The loop is detected, however the error message talks about directory `other` that doesn't exist. This was because of a missing call to `Filename.concat` in `Path.follow_symlinks`. I tried to add it but then we hit the other issue that we can't convert back the resulting filename to a path since it is of the form `../other` :(

But in fact the conversion back to a path is not needed, all we want to check is the following property:

*for every path `d` that is an ancestor of `p`, `Path.follow_symlinks d <> Path.follow_symlinks p`*

So it doesn't matter what the result of `Path.follow_symlinks` is. And in fact, we can even use the pair `(device, inode)` of the resulting file, which simplify things even more since all we need to do is call `Unix.stat`. Additionally, it avoids a few syscalls since we can just replace the call to `Path.is_directory` by a call to `Unix.stat`. I implemented that.

**Comment by rgrinberg** [comment]
@diml you might want to use the `-sfh` flags for `ln` so that successive `$ make test` works

**Comment by diml** [comment]
What is `-h`? `ln` on my system doesn't accept it. Currently I do `rm -rf x c* symlink*` at the beginning of the test

**Comment by diml** [comment]
In fact it shouldn't be necessary as the build_system.ml should do that already

**Comment by diml** [comment]
Ok, the tests are passing in travis now

**Pull request merged**